libsolidity/analysis/NameAndTypeResolver.cpp

```diff
@@ -208,18 +208,18 @@ void NameAndTypeResolver::warnVariablesNamedLikeInstructions() const
 	for (auto const& instruction: evmasm::c_instructions)
 	{
 		string const instructionName{boost::algorithm::to_lower_copy(instruction.first)};
-		auto declarations = nameFromCurrentScope(instructionName, true);
-		for (Declaration const* const declaration: declarations)
+		for (Declaration const* const declaration: nameFromCurrentScope(instructionName, true))
 		{
 			solAssert(!!declaration, "");
-			if (dynamic_cast<MagicVariableDeclaration const* const>(declaration))
-				// Don't warn the user for what the user did not.
-				continue;
-			m_errorReporter.warning(
-				8261_error,
-				declaration->location(),
-				"Variable is shadowed in inline assembly by an instruction of the same name"
-			);
+			// Only warn about variables, everything else is not reachable from
+			// inline assembly.
+			if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(declaration))
+				if (!varDecl->isStateVariable() || varDecl->isConstant())
+					m_errorReporter.warning(
+						8261_error,
+						declaration->location(),
+						"Variable is shadowed in inline assembly by an instruction of the same name"
```

Comment:
Just noticed that this warning is missing a fullstop. Will need to update the one test that gets triggered `syntaxTests/inlineAssembly/shadowing/variable_by_opcode.sol`
```suggestion
						"Variable is shadowed in inline assembly by an instruction of the same name."
```